### ISSUE TITLE
Event `pre_update` not updating dirty attributes

### ISSUE THREAD

**Issue by soderluk** [threatify/arango-orm#53]
I'm using an event listener for my models with the  `pre_add` and `pre_update` events. The events update the `updated_at` and `created_at` attributes of the model.
Now the issue is, that when running `db.update(entity, only_dirty=True)`, the update won't catch the modified attribute `updated_at` as dirty, because the event is dispatched AFTER the dirty attributes are checked:
```
        data = entity._dump()

        if only_dirty:
            if not entity._dirty:
                return entity
            data = {k: v for k, v in data.items() if k == '_key' or k in entity._dirty}

        dispatch(entity, 'pre_update', db=self)

        setattr(entity, '_db', self)
        res = collection.update(data, **kwargs)
```
The event should be dispatched BEFORE the `if only_dirty:` so that the update would also update the modified attributes in the event listener.

**Comment by soderluk** [comment]
@kashifpk Any updates on this?

**Comment by soderluk** [comment]
@kashifpk: This is quite a critical issue, and would be nice to hear if there's any progress on this?

**Comment by lietu** [comment]
It might actually be necessary to update that `data` after the event handler for it to work properly, as you don't want to trigger the event handler if there is to be no update.

I.e.

```
        dispatch(entity, 'pre_update', db=self)

        if only_dirty:
            data = {k: v for k, v in data.items() if k == '_key' or k in entity._dirty}
```

**Comment by kashifpk** [comment]
Sorry for the delay in getting back. pre_update is now fires before entity._dirty check.

https://github.com/threatify/arango-orm/commit/b651b6eee75a178ed5b3fffdb4b259894b923528

0.5.6 pushed to pypi too.

For changes to existing code/features, pull requests are also welcome as long as the test cases pass :-)

**Comment by lietu** [comment]
@kashifpk I would say there's now a bug in the implementation, as it triggers `pre_update` events even for non-dirty items that then won't be later updated.

Likely the most efficient solution would be

```
        if only_dirty:
            if not entity._dirty:
                return entity
            dispatch(entity, 'pre_update', db=self)  # In case of updates to fields
            data = {k: v for k, v in data.items() if k == '_key' or k in entity._dirty}
        else:
            dispatch(entity, 'pre_update', db=self)
```

**Comment by kashifpk** [comment]
I assumed that the case was that there were no other fields in the record that changed. Only the updated field would change within the pre_update handler.  Firing the pre_update event after entity._dirty check in this case would never work.

Perhaps I did not understand the use case correctly?

1. If any data on the entity has changed, entity._dirty check will pass and pre_update would fire any way.

2. Previously if entity had not changed but pre_update needed to change fields then that would not work. With current changes it works. 

3. If we update the code to what you've suggested @lietu then we are back to point 1 are we not?

4. Previously pre_update only fired if data had changed. Now a check is needed inside the pre_update handler to see if any data has changed (for normal use cases). For what @soderluk wanted; if data had already changed before getting to pre_update then it would fire without issues. I think his issue was that he wanted to update some fields even if other data had not changed? unless I misunderstood.

**Comment by lietu** [comment]
The issue was as follows

1. A field has changed in the entity
2. The `if only_dirty:` -check gets the list of dirty fields
3. The `pre_update` -event triggers, and changes something else
4. The `data` is not updated to cover the fields changed by `pre_update` and only the previously dirty fields were stored.

More specifically:

 - You update the `name` field of an entity
 - You trigger an update
 - The `pre_update` event fires and updates `updated_at` -field
 - Only the `name` field -change is saved to the database

With the last change I listed it should work correctly always - the fields updated by `pre_update` are taken into consideration, but `pre_update` is not triggered if the `only_dirty` is `True` and nothing has changed in the entity, and the `data` is collected only after `pre_update` makes additional changes.

However, there is indeed still a bug in my last suggestion because the `data = entity._dump()` was above the `pre_update` - something I neglected earlier.

```python
        if only_dirty:
            if not entity._dirty:
                return entity
            dispatch(entity, 'pre_update', db=self)  # In case of updates to fields
            data = {k: v for k, v in entity._dump().items() if k == '_key' or k in entity._dirty}
        else:
            dispatch(entity, 'pre_update', db=self)
            data = entity._dump()
```

Would take care of that case as well - the `data` is always populated after the dispatch.
